src/iterative_ensemble_smoother/experimental.py

```diff
@@ -6,8 +6,241 @@
+    >>> X = np.array([[1, 2, 3],
+    ...               [0, 0, 0],
+    ...               [1, 2, 3],
+    ...               [1, 1, 1],
+    ...               [1, 1, 1],
+    ...               [1, 2, 3]])
+    >>> list(groupby_indices(X))
+    [(array([0, 0, 0]), array([1])), (array([1, 1, 1]), \
+array([3, 4])), (array([1, 2, 3]), array([0, 2, 5]))]
+
+    """
+    assert X.ndim == 2
+
+    # Code was modified from this answer:
+    # https://stackoverflow.com/questions/30003068/how-to-get-a-list-of-all-indices-of-repeated-elements-in-a-numpy-array
+    idx_sort = np.lexsort(X.T[::-1, :], axis=0)
+    sorted_X = X[idx_sort, :]
+    vals, idx_start = np.unique(sorted_X, return_index=True, axis=0)
+    res = np.split(idx_sort, idx_start[1:])
+    yield from zip(vals, res)
+
+
+class AdaptiveESMDA(BaseESMDA):
+    @staticmethod
+    def correlation_threshold(ensemble_size: int) -> float:
+        """Return a number that determines whether a correlation is significant.
+
+        Default threshold taken from luo2022,
+        Continuous Hyper-parameter OPtimization (CHOP) in an ensemble Kalman filter
+        Section 2.3 - Localization in the CHOP problem
+
+        Examples
+        --------
+        >>> AdaptiveESMDA.correlation_threshold(0)
+        1
+        >>> AdaptiveESMDA.correlation_threshold(4)
+        1
+        >>> AdaptiveESMDA.correlation_threshold(9)
+        1
+        >>> AdaptiveESMDA.correlation_threshold(16)
+        0.75
+        >>> AdaptiveESMDA.correlation_threshold(36)
+        0.5
+        """
+        return min(1, max(0, 3 / np.sqrt(ensemble_size)))
+
+    @staticmethod
+    def compute_cross_covariance_multiplier(
+        *,
+        alpha: float,
+        C_D: npt.NDArray[np.double],
+        D: npt.NDArray[np.double],
+        Y: npt.NDArray[np.double],
+    ) -> npt.NDArray[np.double]:
+        """Compute transition matrix T such that:
+
+            X + cov_XY @ transition_matrix
+
+        In the notation of Emerick et al, recall that the update equation is:
+
+            X_posterior = X_prior + cov_XY @ inv(C_DD + alpha * C_D) @ (D - Y)
+
+        This function computes a transition matrix T, defined as:
+
+            T := inv(C_DD + alpha * C_D) @ (D - Y)
+
+        Note that this might not be the most efficient way to compute the update,
+        since cov_XY := center(X) @ center(Y) / (N - 1), and we can avoid creating
+        this (huge) covariance matrix by performing multiplications right-to-left:
+
+            cov_XY @ inv(C_DD + alpha * C_D) @ (D - Y)
+            center(X) @ center(Y) / (N - 1) @ inv(C_DD + alpha * C_D) @ (D - Y)
+            center(X) @ [center(Y) / (N - 1) @ inv(C_DD + alpha * C_D) @ (D - Y)]
+
+        The advantage of forming T instead is that if we already have computed
+        cov_XY, then we can apply the same T to a reduced number of rows (parameters)
+        in cov_XY.
+        """
+
+        # TODO: This is re-computed in each call. Can we pre compute it?
```

Comment:
I think the current implementation does this outside the loop over parameter groups:

https://github.com/equinor/ert/blob/6aa2aaf5d2a0932c740f95ca1143a98c6460f620/src/ert/analysis/_es_update.py#L544-L550

Does this mean that we would recompute this for ever batch of parameters?

Reply:
Currently yes. I will look into fixing this before merging! Should be an easy fix.